ecommerce/views_test.py

```diff
@@ -1384,6 +1384,54 @@ def test_products_viewset_detail(user_drf_client, coupon_product_ids):
     )
 
 
+def test_products_viewset_expired_programs(user_drf_client):
+    """ Test that the ProductViewSet returns contains only valid programs products and exclude the expired programs correctly"""
+    now = now_in_utc()
+    programs = ProgramFactory.create_batch(3)
+    runs = CourseRunFactory.create_batch(2, course__program=factory.Iterator(programs))
+    index = 0
+    while index < len(programs):
+        ProgramRunFactory.create(
+            program=programs[index],
+            start_date=(now + timedelta(days=1))
+            if index < (len(programs) - 1)
+            else (now - timedelta(days=3)),
+            end_date=(now + timedelta(days=2))
+            if index < (len(programs) - 1)
+            else (now - timedelta(days=1)),
+        )
+        index += 1
+
+    ProductVersionFactory.create_batch(
+        4, product__content_object=factory.Iterator(runs + programs)
+    )
+    response = user_drf_client.get(reverse("products_api-list"))
+    assert response.status_code == status.HTTP_200_OK
+    products = response.json()
+    expired_courseruns = CourseRun.objects.filter(enrollment_end__lt=now).values_list(
+        "id", flat=True
+    )
+    program_ids = [
+        product["latest_version"]["object_id"]
+        for product in products
+        if product["product_type"] == "program"
+    ]
+    # For all the programs in the list there should be on enrollable course run for each associated course
+    assert set(program_ids) == {programs[0].id, programs[1].id}
+    # Expired program should be excluded.
+    assert programs[2].id not in program_ids
+    for program_id in program_ids:
+        program = Program.objects.get(pk=program_id)
+        count = (
+            program.courses.annotate(
+                runs=Count("courseruns", filter=~Q(courseruns__in=expired_courseruns))
+            )
+            .filter(runs=0)
+            .count()
+        )
+        assert count == 0
+
+
```

Comment:
We don't need this in the test since its checking for expired courses and it's already being tested in https://github.com/mitodl/mitxpro/blob/master/ecommerce/views_test.py#L1343-L1353
```suggestion
```

Reply:
I added this just to make it sure again that `program_ids` have the non-expired program ids in list and corresponding count again be 0.